src/cache_cf.cc

```diff
@@ -4777,6 +4782,83 @@ static void free_note(Notes *notes)
     notes->clean();
 }
 
+static DebugMessageId ParseDebugMessageId(const char *value, const char eov, const DebugMessageId oldId)
+{
+    if (oldId > 0)
+        throw TexcHere(ToSBuf("repeated id=... or ids=... option in cache_log_message"));
+    const auto id = xatoui(value, eov);
+    if (!(0 < id && id < DebugMessageCount))
+        throw TexcHere(ToSBuf("unknown cache_log_message ID: ", value));
+    return static_cast<DebugMessageId>(id);
+}
+
+static void parse_cache_log_message(DebugMessages *messages)
+{
+    assert(messages);
+
+    DebugMessage msg;
+    DebugMessageId minId = 0;
+    DebugMessageId maxId = 0;
+
+    char *key, *value;
+    while (ConfigParser::NextKvPair(key, value)) {
+        if (strcmp(key, "id") == 0) {
+            minId = maxId = ParseDebugMessageId(value, '\0', minId);
+        } else if (strcmp(key, "ids") == 0) {
+            const auto dash = strchr(value, '-');
+            if (!dash)
+                throw TexcHere(ToSBuf("malformed cache_log_message ID range: ", key, '=', value));
+            const auto oldId = minId;
+            minId = ParseDebugMessageId(value, '-', oldId);
+            maxId = ParseDebugMessageId(dash+1, '\0', oldId);
+            if (minId > maxId)
+                throw TexcHere(ToSBuf("invalid cache_log_message ID range: ", key, '=', value));
+        } else if (strcmp(key, "level") == 0) {
+            const auto level = xatoi(value);
+            if (level < 0)
+                throw TexcHere(ToSBuf("negative cache_log_message level: ", value));
+            msg.level = level;
+        } else if (strcmp(key, "limit") == 0) {
+            msg.limit = xatoull(value, 10);
```

Comment:
Please disallow repeated same-name options (e.g., cache_log_message id=2 id=22 level=3). I think that is easy to do here using default msg and minId values as canaries.

Reply:
Thank you. I removed the now-bypassed code that protected from repeated id/ids options in commit e4fb372. I forgot that code actually existed when writing my change request, but the request was still valid because there was no protection from repeated level and repeated limit options.